Automattic/jetpack
# Add beta WP Story block

#### Changes proposed in this Pull Request:
This PR introduces a new (Beta) story block for WordPress.com

The idea is that content creation will happen on the Apps, then we need a way to display stories and edit them from WordPress and we figured this could be done via a block living in jetpack.

A few screenshots:

| Story creation | Story rendered as a post |
|---------------|--------------------------|
|<img width="742" alt="Screenshot 2020-07-10 at 11 46 29" src="https://user-images.githubusercontent.com/230230/87141680-b461fd80-c2a3-11ea-8c33-b740b9bfd076.png">|<img width="787" alt="Screenshot 2020-07-10 at 11 47 38" src="https://user-images.githubusercontent.com/230230/87141687-b62bc100-c2a3-11ea-9ce9-202dc3f72d29.png">|

| Story playing  | Story played                      |
|---------------|--------------------------|
|<img width="1287" alt="Screenshot 2020-07-10 at 11 47 59" src="https://user-images.githubusercontent.com/230230/87141695-b7f58480-c2a3-11ea-9976-a52e1a927e1f.png">|<img width="782" alt="Screenshot 2020-07-10 at 11 48 16" src="https://user-images.githubusercontent.com/230230/87141699-ba57de80-c2a3-11ea-9d16-52d4a37c9c1a.png">|

#### Jetpack product discussion
pbD5rk-h
pbD5rk-dS-p2
p1HpG7-9Cp-p2

#### Does this pull request change what data or activity we track or use?
No

#### Review Instructions:

Review as a new block in `extensions/blocks/story`.
The block part and the player part should be reviewed separately as they are quite independent.

The block `edit` function simply mounts the player with some custom configuration (to not play the story in fullscreen for instance).
The block `save` function simply saves the block as an emply div where the player will be mounted. It also includes the list of media files in the block attributes.

Thus rendering the block on the client is fully dynamic. It allows us to evolve Stories while maintaining backward compatibility. In the future we might want to host stories on wordpress.com so they can easily be shared as embeds. In that scenario, the rendering will simply be an iframe.

#### Testing instructions:
- Try creating a story from the block editor (this is not the main flow the user will go through so this is not super polished at the moment). Type "Story", click edit and select Media, try to include a video as well
- View the created story on your site, try to play, pause, mute, unmute, navigate through the slides, replay, exit fullscreen, resize your screen. Try on mobile
- Notice the story renders in full screen on load when displayed as a post

#### Proposed changelog entry for your changes:

* Introduce new experimental "WP Story" block

#### TODO
Here is a list of improvements that might or might not make it for the beta:
- [ ] Handle multiple image ratios: crop top and bottom of the images/videos to avoid showing black margins on the side. If more than X% of the media is hidden start displaying margins on the top and bottom
- [ ] Fix scrolling to the top after story playback
- [ ] Fix video not playing on some mobile browsers or is choppy
- [ ] Other fixes for issues discovered during the pre-alpha (see Issue column https://github.com/orgs/Automattic/projects/191)
